consensus/src/network_interface.rs

```diff
@@ -104,32 +117,50 @@ impl NewNetworkSender for ConsensusNetworkSender {
     ) -> Self {
         Self {
             network_sender: NetworkSender::new(peer_mgr_reqs_tx, connection_reqs_tx),
+            connections: None,
         }
     }
 }
 
 impl ConsensusNetworkSender {
-    /// Send a single message to the destination peer using the `CONSENSUS_DIRECT_SEND_PROTOCOL`
-    /// ProtocolId.
+    /// Send a single message to the destination peer using available ProtocolId.
     pub fn send_to(
         &mut self,
         recipient: PeerId,
         message: ConsensusMsg,
     ) -> Result<(), NetworkError> {
-        let protocol = ProtocolId::ConsensusDirectSend;
-        self.network_sender.send_to(recipient, protocol, message)
+        let remote_protocols = self.connection_metadata(recipient)?;
+        for protocol in DIRECT_SEND {
+            if remote_protocols.contains(*protocol) {
+                return self.network_sender.send_to(recipient, *protocol, message);
+            }
+        }
+        Err(anyhow!("No available protocols for peer {}", recipient).into())
     }
 
-    /// Send a single message to the destination peers using the `CONSENSUS_DIRECT_SEND_PROTOCOL`
-    /// ProtocolId.
+    /// Send a single message to the destination peers using available ProtocolId.
     pub fn send_to_many(
         &mut self,
         recipients: impl Iterator<Item = PeerId>,
         message: ConsensusMsg,
     ) -> Result<(), NetworkError> {
-        let protocol = ProtocolId::ConsensusDirectSend;
-        self.network_sender
-            .send_to_many(recipients, protocol, message)
+        let mut peers_per_protocol: Vec<Vec<PeerId>> = vec![vec![]; DIRECT_SEND.len()];
+        // group by available protocols
+        for r in recipients {
+            if let Ok(remote_protocols) = self.connection_metadata(r) {
+                for (i, peers) in peers_per_protocol.iter_mut().enumerate() {
+                    if remote_protocols.contains(DIRECT_SEND[i]) {
+                        peers.push(r);
+                        break;
+                    }
+                }
+            }
+        }
```

Comment:
Should we log if the recipient doesn't match any protocol?  You could also probably just split out the search of protocols in a separate function, and reuse it between both and collect them, since worst case you have to go through every recipient and every protocol.

Reply:
we probably should log at connection negotiation phase if no protocol matches and reject the connection

Reply:
We do, but I don't think we ensure that specifically Consensus has a protocol, but rather that there is any protocol that matches https://github.com/diem/diem/blob/7bd61c3273be3acb39fa31f11ce261d9c97dbd8e/network/src/protocols/wire/handshake/v1/mod.rs#L224-L247

Reply:
sounds like we should enforce at least one protocol match for each components?

Reply:
We'd basically need a list of applications we expect to support / one of each matches, e.g. public full nodes don't need consensus